### ISSUE TITLE
jcanvas and Browserify

### ISSUE THREAD

**Issue by david-zacharias** [caleb531/jcanvas#201]
I would like to integrate jcanvas into one of our projects using Browserify. Problem is, that jcanvas throws error `Uncaught TypeError: Cannot read property 'PI' of undefined`.

From my point of view the issue is caused by an invalid factory-call `factory( global, true)`. 

Going though the code of jcanvas  v15.02.07:

*Line 8*: In a Browserify-Browser environment the ` if ( typeof module === 'object' && typeof module.exports === 'object' )` resolves to `true`

*Line 9*: Same goes for `global.document`, because `global`is set to `window ` in line 19 (`typeof window !== 'undefined' ? window : this`). 

*Line 10*: The above leads to the call  `factory( global, true )` that sets `window` to `true` leading to `var document` set to `undefined` in *line 24* and so `Math` in *line 28*. 

Finally resulting in the error message.

For me it would work, when jcavas would return the factory-factory
```
function( jQuery, w ) {
    return factory( jQuery, w );
};
```

I am not seeing a use case where `factory( global, true )` could work. So maybe jcanvas can export the  factory-factory whenever `typeof module === 'object' && typeof module.exports === 'object'` is true. 

Looking forward to replies and ideas for solving this!

**Comment by vrbvillor** [comment]
Sorry, I'm not the author of jCanvas, Caleb Evans is.

At 2015-08-11 15:01:25, "David Zacharias" <notifications@github.com> wrote:


I would like to integrate jcanvas into one of our projects using Browserify. Problem is, that jcanvas throws error Uncaught TypeError: Cannot read property 'PI' of undefined.

From my point of view the issue is caused by an invalid factory-call factory( global, true).

Going though the code of jcanvas v15.02.07:

Line 8: In a Browserify-Browser environment the if ( typeof module === 'object' && typeof module.exports === 'object' ) resolves to true

Line 9: Same goes for global.document, because globalis set to window in line 19 (typeof window !== 'undefined' ? window : this).

Line 10: The above leads to the call factory( global, true ) that sets window to true leading to var document set to undefined in line 24 and so Math in line 28.

Finally resulting in the error message.

For me it would work, when jcavas would return the factory-factory

function( jQuery, w ) {
    return factory( jQuery, w );
};


I am not seeing a use case where factory( global, true ) could work. So maybe jcanvas can return the factory-factory whenever typeof module === 'object' && typeof module.exports === 'object' is true.

Looking forward to replies and ideas for solving this!

**Comment by david-zacharias** [comment]
Well, I know. This is why I added my issue here (= in repository of Caleb Evans at https://github.com/caleb531/jcanvas/issues).

**Comment by caleb531** [comment]
Hi, David. I actually didn't write much of the module code—that was contributed by @robertjustjones.

Therefore, @robertjustjones, I would greatly appreciate any insight you could provide on this.

**Comment by robertjustjones** [comment]
David, as #182 mentions, that wrapper was inspired by jQuery since it touches so many different environments and hard to test all of them. Do you have the same issue with jQuery, or can you research any explanation why it works for them?

Also, you might do a PR to show the finished code.  What about a test case?

**Comment by david-zacharias** [comment]
Thanks for your reply!

**jQuery**: jQuery works perfect. I checked their code and they have indeed a similar construct, but they process the parameters in a different way.

**Explanation**: jQuery tests the `true` which is assigned to a var named `nonGlobal` with the statement `if ( typeof noGlobal === strundefined ) {` which makes sense. jcanvas sets variable `window`to `true` and then tries to grab properties from the `true`, which does not make sense.

**Test case**: I can not show the issue in requirebin.com because there is no logging for ecmascript errors. But you can reproduce it by adding these to lines into a file, compile it with browserify, and run the resulting code.
````
var $ = require('jquery');
require('jcanvas');
````

**PR** I am going to do a request today.

**Comment by robertjustjones** [comment]
Sounds good. In working on the above, I used https://github.com/robertjustjones/jcanvas-in-node for testing. You could fork that and use it to test your PR. When you do, you can modify the jcanvas dependency in package.json to point to your PR using the git: syntax. LMK if you need more explanation.

**Comment by david-zacharias** [comment]
I cloned your repository. Making `canvas` run was tricky, but it finally worked. I ran `node index.js` with the current `jcanvas` version (15.2.7) and then with my forked one. Both worked and the resulting base64-encoded images are exactly the same.

**Comment by MyNameIsURL** [comment]
@david-zacharias I changed that one line that you changed in your pull request and my project is still failing at the PI= Math.PI line as well on line 49.  I used npm to install Jcanvas.  So it may be something wrong with the npm repo.

**Comment by david-zacharias** [comment]
@MyNameIsURL With my change on line 49 I did not fix the wrong factory (`factory( global, true )`), but reduced cases in which it is exported. Depending on the variables in your namespace, it can still happen that jCanvas exports the wrong one.

How do you require jcanvas in your code? Do you handover your local `jQuery` and `window` instances like so
```
var $ = require('jquery');
require('jcanvas')($, window);
```

In general: After reading the code I am pretty sure that the factory `factory( global, true )` can not work in any case. But I did not changed it, because I am not sure about which use cases the different setups target.

So lets check thecode. First the parameter going into th IFFE
* `jQuery` is `window.$` whenever `window` is in global namespace, otherwise empty object `{}`
* `global` is `window` whenever `window` is in global namespace, otherwise `this` 
* `factory` is the jCanvas bootstrapping

The code form jCanvas version 15.2.7
 ```javascript
 6: (function( jQuery, global, factory ) {
 7: 
 8:   if ( typeof module === 'object' && typeof module.exports === 'object' ) {
 9:     module.exports = global.document ?
10:       factory( global, true ) :
11:       function( jQuery, w ) {
12:         return factory( jQuery, w );
13:       };
14:   } else {
15:        factory( jQuery, global );
16:   }
```
*Line 8* checks if the code is run in commonJS environment. So if true, jCanvas exports something.
*Line 9* This is where I am not shure why `window.document` is checked.
*Line 10* This factory call will lead to the Math.Pi is undefined issue in any case.
*Line 11 to 13* Factory that is exported to be able to inject a jQuery and window instance into jCanvas.
*Line 15* Instantly calls the factory with given parameters

So for my taste, it would be okay to export the factory-factory (*line 11 to 13*) whenever we are in commonJS environment, so it is up to the developer to inject jQuery and window instances. But because I do not know what the check form *line 9* targets, I did not do that in my PR.

**Comment by david-zacharias** [comment]
@robertjustjones How do we want to continue? Will you accept my pull request?

**Comment by robertjustjones** [comment]
@david-zacharias sorry for the slow response. @caleb531 this looks good to me. I think you should accept the PR.

**Comment by caleb531** [comment]
Excellent. I'll merge it in and build a new jCanvas release now...

**Comment by caleb531** [comment]
Okay, I've released jCanvas v15.08.17 with the Browserify fix. Thanks again for your assistance, @robertjustjones—it is greatly appreciated.

Caleb

**Comment by david-zacharias** [comment]
Nice. Thanks!
